opentracing-spring-web/src/main/java/io/opentracing/contrib/spring/web/client/WebClientSpanDecorator.java

```diff
@@ -0,0 +1,109 @@
+ * the License.
+ */
+package io.opentracing.contrib.spring.web.client;
+
+import io.opentracing.Span;
+import io.opentracing.tag.Tags;
+import org.springframework.web.reactive.function.client.ClientRequest;
+import org.springframework.web.reactive.function.client.ClientResponse;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Decorate span by adding tags/logs or operation name change.
+ *
+ * <p>Do not finish span (otherwise some tags/logs might be missing)!
+ *
+ * @author Csaba Kos
+ */
+public interface WebClientSpanDecorator {
+
+    /**
+     * Decorate span before before request is executed, e.g. before {@code .onSubscribe()} is called.
+     *
+     * @param clientRequest request
+     * @param span client span
+     */
+    void onRequest(ClientRequest clientRequest, Span span);
+
+    /**
+     * Decorate span after the clientRequest is done, e.g. after {@code .onNext()} is called.
+     *
+     * @param clientRequest clientRequest
+     * @param clientResponse clientResponse
+     * @param span span
+     */
+    void onResponse(ClientRequest clientRequest, ClientResponse clientResponse, Span span);
+
+    /**
+     * Decorate span when exception is thrown during clientRequest processing, e.g. when {@code .onError()} is called.
+     *
+     * @param clientRequest clientRequest
+     * @param throwable exception
+     * @param span span
+     */
+    void onError(ClientRequest clientRequest, Throwable throwable, Span span);
+
+    /**
+     * Decorate span when the subscription is cancelled.
+     *
+     * @param clientRequest clientRequest
+     * @param span span
+     */
+    void onCancel(ClientRequest clientRequest, Span span);
+
+    /**
+     * This decorator adds set of standard tags to the span.
+     */
+    class StandardTags implements WebClientSpanDecorator {
+        static final String COMPONENT_NAME = "WebClient/java-spring-web";
+
+        @Override
+        public void onRequest(final ClientRequest clientRequest, final Span span) {
+            Tags.COMPONENT.set(span, COMPONENT_NAME);
+            Tags.HTTP_URL.set(span, clientRequest.url().toString());
+            Tags.HTTP_METHOD.set(span, clientRequest.method().toString());
+
+            if (clientRequest.url().getPort() != -1) {
+                Tags.PEER_PORT.set(span, clientRequest.url().getPort());
+            }
+        }
+
+        @Override
+        public void onResponse(final ClientRequest clientRequest, final ClientResponse clientResponse, final Span span) {
+            Tags.HTTP_STATUS.set(span, clientResponse.rawStatusCode());
+        }
+
+        @Override
+        public void onError(final ClientRequest clientRequest, final Throwable throwable, final Span span) {
+            Tags.ERROR.set(span, Boolean.TRUE);
+            span.log(errorLogs(throwable));
+        }
+
+        @Override
+        public void onCancel(final ClientRequest httpRequest, final Span span) {
```

Comment:
When does cancel happen?

Reply:
Excellent question! It looks like cancellation can happen in various circumstances:
- If the user of the WebClient "unsubscribes" from the request (for example because they have a timeout that they reached and they no longer wish to wait for the result). I think in this case I think it is correct to log an error on the span, since timeout can be considered an error.
- In various Reactor constructs, `cancel()` is called even when the `Mono` (successfully) emits an item. An example of this is [`Flux#next()`](https://github.com/reactor/reactor-core/blob/7bf86a81734bf1dc1a468a99dfdad2590c089de9/reactor-core/src/main/java/reactor/core/publisher/MonoNext.java#L75). Since this is part of the normal (successful) behavior, I don't think we should log an error.

There does not seem to be a good way to differentiate between the above two cases in the tracing code, so I changed this to log a `cancelled` event instead of an error. It will be up to the user who is looking at the trace logs to determine if the cancellation indicates an error or not. (For example the user can infer this from the HTTP status code, etc.)

Reply:
If it is a part of normal execution flow and it happes often we can avoid adding meta-data at all.

Reply:
As an application developer, it is still valuable information that the subscription was cancelled. Cancellation does not happen in simple WebClient flows, only if you're doing something more complex (like retries, or firing off multiple requests and waiting for the fastest one to return).

I think it's worth have these events in the trace logs.

If other people find them bothersome, they can file an issue and we can make it optional, but I think that won't be the case.